File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null)
            return list2;
        if (list2 == null)
            return list1;

        ListNode node = new ListNode();

        if (list1.val < list2.val) {
            node.val = list1.val;
            node.next = mergeTwoLists(list1.next, list2);
        } else {
            node.val = list2.val;
            node.next = mergeTwoLists(list1, list2.next);
        }

        return node;
    }

    public ListNode mergeKLists(ListNode[] lists) {
        int n = lists.length;

        if (n == 0)
            return null;
        if (n == 1)
            return lists[0];

        ListNode newHead = mergeTwoLists(lists[0], lists[1]);

        if (n == 2)
            return newHead;

        for (int i = 2; i < n; i++) {
            newHead = mergeTwoLists(newHead, lists[i]);
        }

        return newHead;
    }
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int n = lists.size();

        if (n == 0)
            return NULL;

        priority_queue<pair<int, ListNode*>, vector<pair<int, ListNode*>>,
                       greater<pair<int, ListNode*>>>
            pq;

        for (int i = 0; i < n; i++) {
            if (lists[i] != NULL) {
                pq.push({lists[i]->val, lists[i]});
            }
        }

        ListNode* dummy = new ListNode(-1);
        ListNode* temp = dummy;

        while (!pq.empty()) {
            pair<int, ListNode*> minNode = pq.top();
            pq.pop();

            temp->next = minNode.second;
            temp = temp->next;

            if (minNode.second->next != NULL)
                pq.push({minNode.second->next->val, minNode.second->next});
        }

        return dummy->next;
    }
};